File: manual/c/globals.c
#include "globals.h"

/*
 * Global variables
 */

struct globalConfig globalConfig = {5, 8};

/*
 * Constants
 */

const int globalConstant = 1;

/*
 * Constant examples
 */

const int constArray1 [4] = {1, 2, 3, 4};
const int constArray2 [] = {1, 2, 3, 4, 5};

const struct tuple constTuple = { 3, 7};
struct tuple nonConstTuple = { 3, 5};

int Int = 8;
const int constInt = 9;
int * ptrToInt = &Int;
const int * ptrToConstInt = &constInt;
int * const constPtrToInt = &Int; // cheating with a cast
const int * const constPtrToConstInt = &constInt;

/*
 * Guidelines for binding generation
 */

int a = 17;
const int a2 = 18;

int * b = &a;
const int * b2 = &a;

triplet c = {19, 20, 21};
const triplet c2 = {22, 23, 24};

typedef int list[];
list d = {25, 26, 27, 28};
const list d2 = {29, 30};

File: manual/hs/manual/app/Manual/Globals.hs
{-# LANGUAGE OverloadedRecordDot #-}

module Manual.Globals (examples) where

import Foreign as F

import HsBindgen.Runtime.IncompleteArray qualified as IA
import HsBindgen.Runtime.PtrConst qualified as PtrConst

import Globals qualified
import Globals.Global qualified as Globals
import Manual.Tools

{-------------------------------------------------------------------------------
  Examples
-------------------------------------------------------------------------------}

examples :: IO ()
examples = do
    section "Global variables"
    config <- peek Globals.globalConfig
    print config
    poke Globals.globalConfig $
      config{Globals.globalConfig_numThreads = 3}
    config' <- peek Globals.globalConfig
    print config'

    subsection "Non-extern globals"
    print =<< peek Globals.nonExternGlobalInt

    subsection "Constants"
    print Globals.globalConstant
    print Globals.anotherGlobalConstant

    subsubsection "Constant examples"
    print Globals.constArray1
    print =<< IA.peekArray 5 (PtrConst.unsafeToPtr Globals.constArray2)
    print Globals.constTuple
    print =<< F.peek Globals.nonConstTuple
    print =<< F.peek Globals.int
    print Globals.constInt
    print =<< F.peek Globals.ptrToInt
    print =<< F.peek Globals.ptrToConstInt
    print Globals.constPtrToInt
    print Globals.constPtrToConstInt
